### ISSUE TITLE
Client Performance Not Great OOTB

### ISSUE THREAD

**Issue by jamesward** [zio/zio-http#2117]
**Describe the bug**
Setting up a `Client` has high overhead ~2 seconds and without connection pooling, there is ~300ms of overhead on each request.  It'd be great to have better out-of-the-box performance.

**To Reproduce**
Project to test performance: https://github.com/jamesward/zio-http-client-perf

On my connection / machine here are some results compared against the JDK's `java.net.http.HttpClient` hitting `https://google.com`

`java.net.http.HttpClient`
 - 1 req = 1375ms
 - 10 sequential = 708ms / req

`Client.default`
 - 1 req = 3038ms
 - 10 sequential = 896ms / req

`FixedConnectionPool size 10`
 - 1 req = 3105ms
 - 10 sequential = 560ms / req

**Expected behaviour**
Lower `Client` creation overhead and lower client reuse overhead (perhaps `Client.default` should use a connection pool)?

**Comment by jdegoes** [comment]
To be specific, performance goals of this ticket should be:

 - Almost no overhead with connection pooling (e.g. connection right away and spin up connection pool in the background)
 - Achieve performance better than `java.net.http.HttpClient` for both single request, and 10 sequential requests (currently the first is 2x slower, while the latter is 200ms slower)

Essentially, out of the box experience should be measurably better than what one could obtain using `java.net.http.HttpClient`, for both single-use scenarios, and batch / high-volume scenarios.

/bounty $600

**Comment by jamesward** [comment]
Yup! Thanks for the clarified goals.

**Comment by tachyonicbytes** [comment]
Hi! I would like to work on this!

I saw that comments on newer PRs say that I should type /attempt #2217 in order to start, but the older ones don't. Should I do it anyway?

**Comment by adamgfraser** [comment]
@tachyonicbytes I think either way is fine! Most important thing is just to let other people know you are working on it which you already have! 😃

**Comment by tachyonicbytes** [comment]
Yeah, algora-pbc[bot] already gave me the rocket emoji, so I guess I did what it wanted

**Comment by dkovalenko** [comment]
We are comparing such a different implementations. Do we want to have something like `.build()` from java HttpClient  on ZClient? In this case we could do some stuff like preparing thread pools and other heavy initialization eagerly.
Also I see that default `Client.config` pool of 10.

<img width="524" alt="image" src="https://github.com/zio/zio-http/assets/874918/c242b3a3-20ec-49a5-abc0-6d1b5ca715ec">


If we compare default vs `.withDisabledConnectionPool` I got that:
```
 1 http java = 383ms / req
 1 http zio = 2937ms / req
 1 http withDisabledConnectionPool = 2224ms / req
```

**Comment by lackhoa** [comment]
Is anyone still working on this? If not then I'll take it!
/attempt #2117

<details id="algora-options">
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clgw19t3v001mma0ffryz0ahd/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by lackhoa** [comment]
@jdegoes Hi, I want to gain more context on this:
> Essentially, out of the box experience should be measurably better than what one could obtain using java.net.http.HttpClient

Why do you think we should be measurably faster (And by how much in your estimate?). Where does the win come from :)?

**Comment by abebeos** [comment]
/attempt #2117

> performance goals

looks like those requirements/goals are not feasible. You need to revise.

**Comment by abebeos** [comment]
> Note: The user @lackhoa is already attempting to complete issue #2117 and claim the bounty. If you attempt to complete the same issue, there is a chance that @lackhoa will complete the issue first, and be awarded the bounty. We recommend discussing with @lackhoa and potentially collaborating on the same solution versus creating an alternate solution.

This is nonsense. Bountysystems should stay out of the bounty-workers way. Gitcoin was a good example of "how to not do it".

**Comment by abebeos** [comment]
@jdegoes , I have the first decent numbers on test-runs, but I've a problem with your bounty-process.

https://console.algora.io/org/ZIO/bounties?status=open does not even list this bounty, and then I'm listed with a red dot here:

https://github.com/zio/zio-http/issues/2117#issuecomment-1521423916

**Comment by abebeos** [comment]
I've essentially solved this issue, but not sure how to proceed. Problems:

* silent team
* verbose bounty bot with a deficient/misleading/annoying process

**Comment by jdegoes** [comment]
@abebeos What problems are you having with the bot?

**Comment by abebeos** [comment]
bot is just the 'messanfer'. Algoras process is flawed. Me nearly ready, but it's unclear to me if i could claim the bounty.

**Comment by lackhoa** [comment]
Hi @jdegoes, I've reproduced the benchmark in the ticket. The main problem with the original benchmark is that it includes the client shutdown time, for example:

```
req(urlHttps).provide(Client.default).timed
```

Our client does take a bit of time to shutdown, but that shouldn't matter in practice because it should only be done at the end of the app anyway. Plus we've reduced that time in #2410, so it shouldn't take 2s anymore.

Here's [my updated benchmark](https://gist.github.com/lackhoa/cd00ab85b99c4d50f8dfcb9072616a6a), which only takes into account client *creation* plus *request* time:

Sample output

```
[info] --- Iteration: 0 ---
[info] No connection pool: {single (1 requests): 844 ms/req}; batch (10 requests): 44 ms/req
[info] Java HttpClient: {single (1 requests): 253 ms/req}; batch (10 requests): 40 ms/req
[info] Default Client: {single (1 requests): 100 ms/req}; batch (10 requests): 23 ms/req

...

[info] --- Iteration: 108 ---
[info] No connection pool: {single (109 requests): 46 ms/req}; batch (1090 requests): 37 ms/req
[info] Java HttpClient: {single (109 requests): 49 ms/req}; batch (1090 requests): 37 ms/req
[info] Default Client: {single (109 requests): 39 ms/req}; batch (1090 requests): 21 ms/req
```

So our single request time is already better than the Java client. But as you can see, the very first request is somehow very slow. I have no idea why, because all subsequent requests are fast (even if you change the host, so it doesn't have anything to do with connection pooling or DNS).

I've also looked at ZKeyedPool and ZPool, and the connections are already done in the background already.

Anyway in conclusion, I think we still have a problem with start-up time, but IMO it's a much smaller problem (because it's only a one-time cost, not *per request*). So I'll stash this one and come back when there's more input.

**Comment by abebeos** [comment]
@jdegoes , the problems here are mostly communicative and in the requirements. @lackhoa 's reply/work essentially adds some more misunderstandings to the overall issue.

After several trial&errors, this came out:

* Issue here is about **"Setting up a Client has high overhead "**
  * Issue is not(!) about the raw request-timing (javaHttp vs Netty)
    * This timing is fine (own tests, results of other devs)

### Problems

* Original reproducer (from @jamesward) is outdated (e.g. does not use the latest version, )
* Reproducer from @lackhoa is different, introduces other problems and some complexity.
* ZIO framework does some 'magic' behind the scenes that changes results (instantiation of client, lifetime of client)

### Solution

* setup a simple test repo based on 
  * https://github.com/jamesward/zio-http-client-perf
  * keep it a simple as possible, update it to latest versions

Repo should allow easy work with a profiler (possible future bounty to fine-tune the internals).

### Bounty

Please let me know how to proceed here, I do not want to interfere with anyones work, but it is clear to me that this needs more attention from someone who focuses on clairty/simplicity (so others can pickup more detailed work easily).

### Cli Example

This simulates a cli single-request. The time for the request is including client-startup. Total time includes overall ZIO startup/shutdown, including client and the request. It shows that for the use-case "single-request via cli", optimization is needed.

```sh
========== Client.default ==========

https://google.com
1st   = 1011 ms/req

TOTAL = 3801ms

========== clientFast ==========

https://google.com
1st   = 1332 ms/req

TOTAL = 3467ms

Process finished with exit code 0
```

(this is with 3.0.0-RC2, latest version needs more work on my side, so the bounty needs to be clarified, or e.g. an additional bounty filed. I could alternatively share the bounty, but the budget is already quite low for the overall effort.)

**Comment by lackhoa** [comment]
@abebeos What misunderstanding did I add to the issue? Or do I have to pay thousands of dollars to find out the answer?

**Comment by abebeos** [comment]
> What misunderstanding did I add to the issue? 
> Or do I have to pay thousands of dollars to find out the answer?

$0 (as everything here I've done so far).

Main misunderstanding:

>>> Anyway in conclusion, I think we still have a problem with start-up time, but IMO it's a much smaller problem (because it's only a one-time cost, not per request). 

this is not "a smaller problem" but the main issue here.

**Comment by lackhoa** [comment]
No it is clearly not the main issue. Why do you think that it is?

**Comment by lackhoa** [comment]
> > No it is clearly not the main issue. Why do you think that it is?
> 
> ... (no further comment)

Ok, if you don’t want to talk, please do me a favor and never @ mention my name again. I don’t want to interact with you ever. You don’t say anything interesting anyway.

**Comment by abebeos** [comment]
lackhoa, all good.
@jamesward , please see https://github.com/jamesward/zio-http-client-perf/issues/1.

**Comment by jamesward** [comment]
I've updated my repo to the latest deps and re-run the tests which produce these results:
```
1 https curl client = 636ms / req
10 https curl client = 537ms / req
1 https jdk client = 1006ms / req
10 https jdk client = 155ms / req
1 https netty epoll client = 359ms / req
10 https netty epoll client = 578ms / req
1 https netty nio client = 242ms / req
10 https netty nio client = 533ms / req
1 http zio = 2879ms / req
10 http zio = 457ms / req
1 https zio = 2738ms / req
10 https zio = 348ms / req
1 https zio fixed pool = 2742ms / req
10 https zio fixed pool = 372ms / req
```

**Comment by abebeos** [comment]
very nice, see https://github.com/jamesward/zio-http-client-perf/pull/2 (using SNAPSHOT would be next step, tried, but sbt makes it not that trivial as maven). After switching to the nio-htttp-3.0.0-SNAPSHOT, trial&error tests with different configurations can already determine better speed.

The main thing is to have the repo in a condition where anyone can start immidiately with the detail-level code-work (on config&nio-http).

**Comment by jamesward** [comment]
Based on a suggestion from @abebeos I've bumped my test to a recent zio-http SNAPSHOT and for OOTB usage I'm seeing significantly better performance!
```
# Baseline
1 https jdk client = 451ms / req
10 https jdk client = 31ms / req

# 3.0.0-RC2
1 http zio = 2410ms / req
10 http zio = 266ms / req
1 https zio = 2352ms / req
10 https zio = 256ms / req
1 https zio fixed pool = 2186ms / req
10 https zio fixed pool = 255ms / req

# 3.0.0-RC2+73-6411c97a-SNAPSHOT
1 http zio = 666ms / req
10 http zio = 74ms / req
1 https zio = 237ms / req
10 https zio = 64ms / req
1 https zio fixed pool = 2160ms / req
10 https zio fixed pool = 350ms / req
```

Those initial numbers are looking much better for the OOTB usage.  Not sure what changed / fixed this since RC2.

**Comment by abebeos** [comment]
> 1 http zio = 666ms / req
> 10 http zio = 74ms / req
> 1 https zio = 237ms / req
> 10 https zio = 64ms / req

those numbers look wrong, looking at it now.

**Comment by abebeos** [comment]
please see https://github.com/jamesward/zio-http-client-perf/issues/5

**Comment by abebeos** [comment]
my numbers for cli-like invocation:

```
// default contains pool and fast shutdown, this has changed since the original issue was filed)

ZIO default, 1 request (https://google.com)
1st   = 1073 ms
TOTAL = 2218ms

ZIO fast-shutdown, no-pool, 1 request (https://google.com)
1st   = 1032 ms
TOTAL = 1851ms
```

So, currently, the fastests **curl-like single-request** would be close to 2 seconds (vs half a second for curl). Batch requests are fine, if pooling is enabled (default client).

I'll update my repo (fork of jameswards, that contains non-complex code, thus choosen by me) later the day to make those reproducible.

Further efforts (and further bunties) should be focused on:

* Increase curl-like single-request execution speed
  * possibly: focus on "time to send the request" (shutdown time is secondary)

**Comment by abebeos** [comment]
This should do it, easy to understand code, most relevant code on top:

https://github.com/abebeos/zio-http-client-perf/blob/main/src/main/scala/App.scala

To get even more precise results, this can happen:

- https://github.com/jamesward/zio-http-client-perf/issues/5

People who know the ZIO internals could now work e.g. on:

* Client.cli
* Scope.cli

where 'cli' implies "minimum overhead for cli invocation"

**Comment by abebeos** [comment]
@jdegoes , you'll find my results re this issue/bounty here:

- https://github.com/abebeos/zio-http/issues/1

Algora's bounty process focuses on code & PR, mid-level devs. 

Relevant pre-work (e.g. requirements/feasibility analysis, clarifying status, fixing bugs in reproducers, making the domain-knowledge more accessible) cannot be send it via a PR. Management work cannot be send in via a PR.

I would just write some more documentation before I would claim the bounty here (the $600), and as for a followup bounty to go on with enhancing the speed more (e.g. $1000 per 10%).


Several follow-up issues:

* Make contribution/bounty-processes more flexible
* Ensure that the initial issue-text is updated as the issue evolves
* Ensure that the documentation is online for the "main" branch (e.g. "version: snapshot")
* etc.

**Comment by adamgfraser** [comment]
@abebeos The bounty is for actually making the performance faster than `java.net.http.HttpClient` in both single use and batch scenarios, as clearly stated above. You have not done that but have merely restated the issue. Creating benchmarks and identifying drivers is something I would expect to be a part of any PR that addresses this issue.

**Comment by ghost** [comment]
> The bounty is for actually making the performance faster than `java.net.http.HttpClient` in both single use and batch scenarios

Please be aware that you are delusional if you believe that this is feasible.

Even a $60_000 bounty could not achieve this result.

> You have not done that but have merely restated the issue.

"merely restated"... 

Wow!

Best case scenario: you (@adamgfraser & @jdegoes ) are just to busy to take time to understand a problem-domain.

I rest my case and leave it to the code-monkeys that consume anything served as-is!

Cu around.

@jamesward I may still keep a look on your benchmark, mostly out of curiosity (e.g. how a simple test-runner would be implemented in scala, https://github.com/jamesward/zio-http-client-perf/issues/5#issuecomment-1754947484)

**Comment by abebeos** [comment]
(flushed account and back)

**Comment by abebeos** [comment]
related: https://github.com/zio/zio-http/issues/2265

**Comment by lazaridiscom** [comment]
> The bounty is up for grabs! Everyone is welcome to `/attempt #2117` 🙌

Algora, you are a joke.

**Comment by abebeos** [comment]
> The bounty is up for grabs! Everyone is welcome to `/attempt #2117` 🙌

Oh my...

**Comment by abebeos** [comment]
> @abebeos: Reminder that in 7 days the bounty will become up for grabs, so please submit a pull request before then 🙏

Can you stop this bot from spamming? Enough is enough.
